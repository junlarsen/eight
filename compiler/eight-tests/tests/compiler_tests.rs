//! End-to-end tests for the compiler with a variety of different inputs and compiler options.

macro_rules! compiler_test_suite {
    ($name:ident, $directory:expr, [$($arg:expr),*]) => {
        #[test]
        fn $name() {
            use std::process::Command;

            insta::with_settings!({
                description => format!("compiler test suite for {}", stringify!($name)),
                snapshot_path => concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/", stringify!($name)),
            }, {
                insta::glob!(concat!($directory, "/**/*.test"), |path| {
                    let mut cmd = Command::new(insta_cmd::get_cargo_bin("eightc"));
                    $(cmd.arg($arg);)*
                    let relative = path.strip_prefix(env!("CARGO_MANIFEST_DIR"))
                        .expect("failed to build relative path from Cargo.toml");
                    insta_cmd::assert_cmd_snapshot!(cmd.arg(relative));
                });
            })
        }
    }
}

compiler_test_suite!(ui, "ui", []);
compiler_test_suite!(syntax, "syntax", ["--emit-ast"]);
compiler_test_suite!(syntax_lowering, "syntax-lowering", ["--emit-hir"]);
compiler_test_suite!(type_inference, "type-inference", ["--emit-hir"]);
