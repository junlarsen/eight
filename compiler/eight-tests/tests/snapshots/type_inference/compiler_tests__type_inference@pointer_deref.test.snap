---
source: compiler/eight-tests/tests/compiler_tests.rs
description: compiler test suite for type_inference
info:
  program: eightc
  args:
    - "--emit-hir"
    - tests/type-inference/pointer_deref.test
input_file: compiler/eight-tests/tests/type-inference/pointer_deref.test
snapshot_kind: text
---
success: true
exit_code: 0
----- stdout -----
hir_module {
  declares {
    // module types
    
    
    // module struct types
    
    
    // module functions
    fn address_of(a: i32) -> unit;
    fn deref(a: i32, b: **i32) -> unit;
    
    // module traits
    
    
    // module instances
  
  }
  
  defining {
    // module functions
    fn address_of<>(a: i32) -> unit {
      let k: *i32 = (&(1 as i32) as *i32);
    }
    fn deref<>(a: i32, b: **i32) -> unit {
      let m: i32 = (*(*(b as **i32) as *i32) as i32);
    }
    // module intrinsic functions
    
  
  }
}

----- stderr -----
